### PR TITLE
Proof of concept DB-driven approach.

### PR DESCRIPTION
**Pull request by cowboy**
I'm using [ngrok](https://ngrok.com/) to test this. Here's how you can test it:

1. Install ngrok.
1. Run `ngrok http 8000` in a separate window. This process will need to stay running for a while.
1. Note the HTTPS url, it will look something like <https://e2d1dd50.ngrok.io>.
1. Create a new Slack app at <https://api.slack.com/apps/new>
  1. Enter an app name like "Pombot Cowboy Test" (but using your name)
  1. Enter placeholder short and long descriptions like "lorem ipsum"
  1. Enter the HTTPS url from step 3 in "Link to clear instructions"
  1. Click "Create app."
1. Click on "App Credentials" in the left nav.
  1. Create a `.env` file in the project root that looks like the example below, using the Client ID and Client Secret from this page.
  1. Enter the HTTPS url with `/authorize` at the end into Redirect URI(s). It'll look something like <https://e2d1dd50.ngrok.io/authorize>
  1. Click "Save Changes."
1. Click on "Bot Users" in the left nav.
  1. Click the "Add a bot to this app" button.
1. Run the app server with `node app.js`
1. Visit the HTTPS url in your browser.
1. Click the "Add to Slack" button.
1. Select Bocoup.
1. Note the bot name and click the "Authorize" button.
  * You should see the node process create the bot (the name shown will be the one hard-coded into app.js)
  * You should be able to interact with the bot in DM, like `/dm @your_bot_name hello world`

Sample .env file:

```bash
SLACK_CLIENT_ID=1234567890
SLACK_CLIENT_SECRET=1234567890abcdef
```

### Notes

* You should be able to kill the node app process and restart it, and the bot should restart automatically with the process.
* You should be able to find the bot in the list at <https://bocoup.slack.com/apps/manage> and click the "Remove App" button and see an eventual "account_inactive" error and the bot being deleted.
* You should be able to kill the node app process and restart it, and only non-deleted bots should restart automatically with the process.
* Multiple users should be able to authorize the same app (via the same HTTPS url) simultaneously, and only 1 bot instance should be running at a time.

### CODE REVIEW

**Comment by tkellen** [comment]
I think we should use [cluster](https://nodejs.org/api/cluster.html) for this rather than spinning up all of our instances in a single node process.

File: app.js [hunk 1]
```diff
@@ -1,45 +1,176 @@
...
+db.run(`
+  CREATE TABLE IF NOT EXISTS bots (
+    id SERIAL PRIMARY KEY,
+    token TEXT,
+    is_active BOOLEAN NOT NULL DEFAULT false
+  )
+`);
+
+function addBotIntegration(token, cb) {
+  db.get("SELECT id FROM bots where token = ?", token, (err, id) => {
```
Comment (tkellen) [inline app.js]: Once we move to PostgreSQL, let's use "upsert" functionality so we can do this in a single round trip to the DB. This feature recently landed in 9.5 and I *just* [flagged our RDS instance to upgrade to it](https://github.com/bocoup/infrastructure/commit/8013269a6eb92102606c19ef56cf50b8474ff2af)

```sql
INSERT INTO bots (token)
VALUES ('foo')
ON CONFLICT (token)
DO UPDATE SET is_active = true
```

File: app.js [hunk 2]
```diff
@@ -1,45 +1,176 @@
...
+  bot.login();
+  return bot;
+}
+
+const db = new sqlite3.Database('bots.db');
 
+db.run(`
+  CREATE TABLE IF NOT EXISTS bots (
+    id SERIAL PRIMARY KEY,
+    token TEXT,
```
Comment (tkellen) [inline app.js]: we should enforce uniqueness on the token field.

Comment (tkellen) [inline app.js]: In PostgreSQL you can accomplish that by adding `UNIQUE` after the column definition. You can also do it outside the table definition like so:
```sql
ALTER TABLE bots ADD CONSTRAINT bot_token_key UNIQUE (token);
```

Before you ask, constraints live in a global namespace, so prefixing with bot_ is needed if you want to follow the convention that PostgreSQL itself uses.

File: app.js (hunk 1 above)
Comment (cowboy) [inline app.js]: Ok, I googled for "upsert" and found a whole bunch of confusing stuff, nothing like what you just posted. Awesome!

File: app.js (hunk 2 above)
Comment (tkellen) [inline app.js]: PS: there is a rubric of naming conventions for postgresql here:
http://stackoverflow.com/questions/4107915/postgresql-default-constraint-names

**Comment by cowboy** [comment]
> I think we should use cluster for this rather than spinning up all of our instances in a single node process.

As far as I can tell, cluster just gives us one forked child process per cpu. Does it do more than that? IS that why you were suggesting it?

**Pull request closed without merging**